core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala

```diff
@@ -147,18 +151,29 @@ object FileCommitProtocol extends Logging {
       className: String,
       jobId: String,
       outputPath: String,
-      dynamicPartitionOverwrite: Boolean = false): FileCommitProtocol = {
+      dynamicPartitionOverwrite: Boolean = false,
+      isInsertIntoHadoopFsRelation: Boolean = false,
+      isOverwrite: Boolean = false,
+      staticPartitionKVs: Seq[(String, String)] = Seq.empty[(String, String)]):
```

Comment:
Now it's nearly impossible for users to plug in a custom `FileCommitProtocol` as there are so many parameters to take care of. We need to think about what we expect users to plug in.

Reply:
How about encoding the mode of FileCommitProtocol?
We can encode (dynamicPartitionOverwrite, isInsertIntoHadoopFsRelation and isOverwrite) to a parameter, modeCode.
- 0 non-InsertIntoHadoopFsRelation operation 
- 1 dynamicPartitionOverwrite
- 2 staticPartitionOverwrite
- 3 overwrite non-partitioned table
- 4 append

Reply:
I think we may need a FileSourceWriteDesc class here to escape the added parameters here.
And to be compatible with old versions, a new method signature should be added. 
WDYT @cloud-fan and @turboFei ?

Also , the comment in the class doc should be updated correspondingly.

Reply:
@advancedxy 
Thanks for your suggestion. I will update the code, comment and UT.
As mentioned in the comment by you, we should disable concurrent appending to the same partition. 
So we can remove the `isOverwrite` parameter here.

It's a good idea to define a `FileSourceWriteDesc` class to escape `isInsertIntoHadoopFsRelation` and `staticPartitionKVs`. 
Thanks again.

Reply:
Should we escape dynamicPartitionOverwrite in `FileSourceWriteDesc`? It is also a file source write option.